enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,201 @@
+with external load balancers. In Microshift, as there is only one node in the
+cluster, we don't really need to have a load balancer. However, from user
+experience perspective, it's important for customers to be able to migrate
+the workload to Microshift without modification. So we need to find a way to
+support the loadbalancer type Service without a load balancer.
+
+### User Stories
+
+As a Microshift user, I want to reuse the Service manifest that with type of
+`LoadBalancer` without modification. I don't want to allocate a new IP address
+to the Node as the loadbalancer VIP. I want to use the NodeIP as the ingress IP
+of the loadbalancer Service. 
+
+### Goals
+
+* Allow users to deploy the load balancer type Service to Microshift.
+* The node IP shall be used as the ingress IP of the Service.
+* As we will support pluggable CNI in Microshift. The implementation shall be
+also compatible with CNIs other than OVN-Kubernetes.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+To support the loadbalancer type Service, we need to create a new controller
+component that watch the Service objects. It shall be responsible for:
+
+1. work with CNIs to plumb the ingress traffic to the endpoint pods.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+### Workflow Description
+
+**cluster user** is a human user responsible for provisioning Services to a
+Microshift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+This new controller is part of the microshift binary. It shall be able to know
+what CNI is used by the cluster, and behave differently based on the CNI type.
+
+### Risks and Mitigations
+
+The pluggable CNI design of Microshift is still under discussion, we need to
+make sure that CNI type can be aware in Microshift binary.
+
+### Drawbacks
+
+It is not a CNI agnostic design. The loadbalancer Service controller will behave
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+However, in the context of Microshift, as we only get one node in the cluster,
+it doesn't make sense to use a real load balancer to forward the ingress
+traffic. So instead of provisioning the loadbalancer instance, the controller
+is only responsible for helping the CNIs to plumbing the ingress traffic.
+
+### Traffic Plumbing
```

Comment:
If we say that we do not want to bring klipper into the product, how would you approach the division in behavior described here? To me it seems like the load balancer implementation either needs to be able to be swapped out (making it another plugin type) or we need a configuration option to control the built-in implementation for cases where OVN-K is not being used. Is that right, or were you thinking of another approach?

Reply:
> If we say that we do not want to bring klipper into the product, how would you approach the division in behavior described here?

The implementation of klipper is quite lightweight. Its image only contains a simple script that configures iptables rules. We can implement the same logic in either microshift binary or hostnetwork pods(the k3s/klipper's approach). The advantage of using hostnetwork pods is that: 1) the pod will also reserve the host port. It takes advantage of the kubernetes scheduling so that it avoids the racing condition; 2) K3S supports multi-nodes, and the pod can be deployed with DaemonSets on each node and configure the node.

From MicroShift perspective, if we don't want to take the klipper approach. We can use microshift binary to control the host iptables rule for most of the non-OVN-K CNIs.

> To me it seems like the load balancer implementation either needs to be able to be swapped out (making it another plugin type) or we need a configuration option to control the built-in implementation for cases where OVN-K is not being used. Is that right, or were you thinking of another approach?

In my assumption, to support other CNIs, we'll have a field in the microshift configuration file to indicate which CNI will be set up for the cluster. The loadbalancer service controller can use this information to determine what it needs to do.

Reply:
Apart from klipper, kube-proxy is another option. Although, it consumes more resources. Some CNIs need to work with it to support the service anyway.

Reply:
@pliurh , can you make sure both of those alternatives are listed at the bottom of the doc? Then we can resolve this thread.

Reply:
Yes, I did.